### DISCUSSION TITLE
How can HR depend on *in-existence* of another HR

### DISCUSSION THREAD

**Discussion by or-shachar** [fluxcd/flux2#3925] [Q&A]
- We have a special case of needing to remove existing HR and create a new one ([root reason is this](https://github.com/fluxcd/flux2/discussions/2854))
- The current `dependsOn` field of HR only allows to depend on existence of another HR.
- Is there an option to depend on the **inexistence** of some HR - to promise clean replacement?

**Comment by kingdonb** [comment]
We are trying to understand the use case for this. I have some hypothesis and we've just been over the scenarios verbally at Bug Scrub, but I'd like to understand *your* use case in particular since you are the one asking for this now :D 

I had the idea that you might be trying to atomically switch from one release-distrib chart to another chart in a seamless way. We have a couple of ideas for how you might work around whatever it is right now, but without knowing the specific hurdle that you're trying to get over, not sure if they will work.

You could try suspending the old HelmRelease, if you're trying to take over and upgrade? Can you shed some light on what it is you're trying to do here, please!

**Reply by or-shachar** [reply]
Hey @kingdonb ! First of all thanks a lot.

Here's my scenario:
- We have a single monorepo with all of our charts and values (in some template language to allow managing dozens of clusters)
- A "job" of updating Flux's cluster states takes ~4 minutes - running the template tool + validation + pushing to the cluster state repo.
- The job is usually part of a bigger workflow that may take 20 minutes.

The scenario where I first suspend/delete the old helmrelease and only after add the new one requires me to do it in two pushes to my monorepo --> two commits to cluster states --> two cycles of workflows --> possible 20 minutes downtime in my service.

The reason I can't delete/suspend the old helmrelease + create the new one in the same commit is that the helm-controller reconciliation loop does not guarantee any order apart of the `depends_on`, so it might choose to first reconcile the new release and then remove/suspend the first one. (which will create a situation that the new release will fail).

**Reply by kingdonb** [reply]
This helps understanding the problem a lot. Thank you for the narrative description.

I'm not aware of any health checking or dependency tool that is oriented towards proving non-existence. That's not to say it isn't a useful idea, but we would want to think through how does this work (design) in order to build it in an optimal way.

But having a coherent problem statement and storyline helps a lot!

Since we are apparently talking about some kind of foot gun (I referenced the linked thread, and I see there's a subtle difference between the behavior of `kubectl apply --force` and `kubectl replace --force`, which is maybe considered as a root cause for this problem, as there is an important lever that is not exposed by Helm upstream.)

I think I would aim for a different approach, since HelmReleases are not quite deterministic, and rollouts can fail for reasons which are unrelated to the one you have already thought to mention, you typically want the new HelmRelease online concurrently with the old one, so you can reuse a load balancer temporarily, or at least confirm the new service is online before shuttering the old one. But there are certainly cases where that isn't possible, like we have a persistent volume that needs to be migrated and it has the ReadWriteOnce semantic. So it's literally impossible to safely begin the installation until after the prior instance of the service shuts off its own deployment or statefulset and breaks its association with the volume and its claim; we need a new claim to reuse the existing volume.

It can certainly be made to work the way you imagine, and it sounds like it could work reliably (with practice) to make it work like you describe. I don't have any ideas at this time for how to do this off the top of my head, without changing Flux itself, which would have to add a special syntax or flag on the existing Kustomization healthChecks or on a HelmRelease dependsOn. It's not obvious to me whether it should go in HelmRelease directly, or if it needs to go in the parent Kustomization so it can be managed in concert with the auxiliary resources like a PV managed separately with its own definition outside of the HelmRelease chart templates.

I think it would go out there in Kustomization, and you'd probably need an extra vestigial Kustomization to get the job done. (It begins to feel like Kustomization controller might not be the tool for this job, but I still see and understand exactly what you are trying to do.)

It might make more sense to me now if I understood why your workflows take 4-20 minutes. If it's just delay from the reconciling intervals and how they might align or fall out of alignment, this delay can be completely mitigated away through the use of Notification controller and the Receiver webhook. GitOps does not need to be on a 5-10 minute lag. You can queue up pull requests and get them pre-approved so they can be applied in order, and the applies can proceed immediately after merging if you use a receiving webhook. It should not be a case of getting hung up in CI, it should be possible to bypass CI unless it's from merge checks that are mandatory. But in most cases where the process is what's getting in the way, I'd probably look to amend the process, rather than adding vestigial features for our tools that only get such an infrequent and obscure use.

I am anticipating that many organizations and network policy structures will provide resistance against setting up Receiving webhooks, I bet that you don't have webhooks, but I do think it's important enough to mention the webhook receivers, since it is officially one part of GitOps GA and the Flux 2.0.0 release announcement that is still forthcoming. In Flux v2.0.0, Flux is marking only 3 APIs as "v1": GitRepository, Kustomization, and Receiver, according to the [roadmap](https://fluxcd.io/roadmap/). There is not null signifigance to the fact that those three APIs are all named in the first GitOps GA milestone. They're all considered important enough to "name first," and also developed enough to consider as stable and permament infrastructure which will always be supported. That's what I think v1 intends, when it comes across our Flux GA news desk.

I do understand that a lot of organizations balk at webhook receivers for various reasons, but I consider that it is a critical part of the pipeline for timely (immediate) changes to flow smoothly in a GitOps delivery pipeline; I would not consider webhook receivers as an optional part if I were in your shoes of trying to coordinate teardown and reinstall with only access through the GitOps pipeline. I think this can make the changes instantaneous and maybe that solves your issue, or maybe you would rather have the footgun, let's talk about it. We see this Receiver structure as particularly important in environments where the admin cannot connect directly to Kubernetes, and so they are not able to run `flux reconcile xxx` on their laptop in order to force a change to go through right away on-demand in a crisis. They shouldn't ever think of this, because each repository has been properly instrumented with webhooks, so unless there is something very wrong with the cluster, changes are always applied right away.

The GA announcement coming up gives us an opening to inquire with our adopters, and some documentation to float up the chain in case of this resistance. I'll pose the question to you, @or-shachar – are you using all 3 of the GA-targeted APIs? Have you incorporated webhook receivers, or are you potentially experiencing pain and problems that stem from that you may have only adopted 2 out of 3 APIs that are planned to be heralded in the "GitOps GA" part 1 milestone?
